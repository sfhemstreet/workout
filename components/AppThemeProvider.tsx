import { ThemeProvider } from "@emotion/react";
import { ReactNode } from "react";

import { darkColors, lightColors } from "../styles/colors";
import { GlobalStyles } from "../styles/GlobalStyles";
import { theme } from "../styles/theme";
import { ThemeType } from "../types/ThemeType";

type AppThemeProviderProps = {
  children: ReactNode;
  themeType: ThemeType;
};
/**
 * AppThemeProvider
 * - provides the current theme in use to child components.
 * - provides the current theme to Emotion styled components via Emotion's ThemeProvider.
 * - applies global CSS via GlobalStyles.
 *
 * @param children
 */
export const AppThemeProvider: React.FC<AppThemeProviderProps> = ({
  children,
  themeType,
}) => (
  <ThemeProvider
    theme={{
      ...theme,
      name: themeType,
      colors: themeType === "LIGHT" ? lightColors : darkColors,
    }}
  >
    <GlobalStyles themeType={themeType} />
    {children}
  </ThemeProvider>
);
