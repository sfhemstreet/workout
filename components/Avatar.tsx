import styled from "@emotion/styled";
import {
  SwitchTransition,
  Transition,
  TransitionStatus,
} from "react-transition-group";
import { FadeInKeyframe } from "../styles/keyframes";

import { SurfaceElevation } from "../styles/SurfaceElevation";
import { FadeInOutMixin } from "./FadeInOut";
import { LoadingShimmer } from "./LoadingShimmer";


type AvatarProps = {
  isLoading: boolean;
  src: string;
  onClick?: () => void;
};

/**
 * Avatar
 * 
 * Displays round user avatar 
 * 
 * @param isLoading when `true` a LoadingShimmer is rendered instead of avatar
 * @param src src of avatar img
 * @param onClick function to call when avatar is clicked
 */
export const Avatar = ({ isLoading, src, onClick }: AvatarProps) => (
  <BtnContainer
    noHover={onClick === undefined}
    onClick={() => !isLoading && onClick && onClick()}
    aria-label="User information"
  >
    <SwitchTransition mode="out-in">
      <Transition key={isLoading ? "true" : "false"} timeout={200}>
        {(status) => (
          <>
            {isLoading ? (
              <LoadingContainer status={status}>
                <LoadingShimmer />
              </LoadingContainer>
            ) : (
              <Img
                role="button"
                status={status}
                src={src}
                alt="User Avatar"
                title="Toggle user menu"
              />
            )}
          </>
        )}
      </Transition>
    </SwitchTransition>
  </BtnContainer>
);

const BtnContainer = styled.button<{ noHover: boolean }>`
  padding: 0px;
  border: none;
  overflow: hidden;

  ${(p) => SurfaceElevation(p.theme.name, 4)};

  width: 50px;
  height: 50px;
  border-radius: 50%;

  display: flex;
  justify-content: space-around;
  align-items: center;

  animation: ${FadeInKeyframe} 300ms linear both;
  transition: ${(p) => p.theme.transitions.normal};

  cursor: ${(p) => (p.noHover ? "default" : "pointer")};

  :hover {
    ${(p) => !p.noHover && `background: ${p.theme.colors.secondary}`};
  }

  @media ${(p) => p.theme.media.tablet} {
    width: 60px;
    height: 60px;
  }
`;

const Img = styled.img<{ status: TransitionStatus }>`
  width: 46px;
  height: 46px;
  border-radius: 50%;

  ${(p) => FadeInOutMixin(p.status)};

  @media ${(p) => p.theme.media.tablet} {
    width: 56px;
    height: 56px;
  }
`;

const LoadingContainer = styled.div<{ status: TransitionStatus }>`
  width: 46px;
  height: 46px;
  border-radius: 50%;

  overflow: hidden;

  display: flex;
  justify-content: space-around;
  align-items: center;

  transition: ${(p) => p.theme.transitions.normal};
  opacity: ${(p) => (p.status === "entered" ? 1 : 0)};

  @media ${(p) => p.theme.media.tablet} {
    width: 56px;
    height: 56px;
  }
`;
