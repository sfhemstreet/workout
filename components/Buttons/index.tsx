export { SubmitButton } from "./SubmitButton";
export { SecondaryButton } from "./SecondaryButton";
export { TertiaryButton } from "./TertiaryButton";
export { Hamburger } from "./Hamburger";
export { SmallRoundButton } from "./SmallRoundButton";
export { ArrowButton } from "./ArrowButton";
export { InlineButton } from "./InlineButton";
export { SelectableInlineButton } from "./SelectableInlineButton";
