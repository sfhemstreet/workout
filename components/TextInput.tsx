import styled from "@emotion/styled";
import { useState } from "react";
import { TransitionStatus } from "react-transition-group";

import { FadeInKeyframe } from "../styles/keyframes";
import { FadeInOut } from "./FadeInOut";
import { HiddenLabel } from "./HiddenLabel";
import { EyeIcon } from "./Icons";

type TextInputProps = {
  id: string;
  value: string;
  onChange: (str: string) => void;
  label: string;
  placeholder: string;
  error?: string;
  list?: string;
  isLabelHidden?: boolean;
  isPassword?: boolean;
  isEmail?: boolean;
  isDisabled?: boolean;
  isOutlined?: boolean;
  isTextArea?: boolean;
  isRequired?: boolean;
  autoComplete?: "new-password" | "off";
  onEnterKeyPress?: () => void;
  tabIndex?: number;
};

/**
 * TextInput
 * 
 * Controlled input for text / password / email, or textarea.
 * 
 * Defaults to `input`
 * Default input type is 'text'
 * 
 * @param id Unique ID for input
 * @param value string value of input
 * @param onChange function called when input change occurs
 * @param label label for input  or textarea
 * @param placeholder placeholder text for input / textarea
 * @param error String error message to display
 * @param list optional datalist id for browser to show a list of options
 * @param isTextArea when `true` renders a textarea instead of an input
 * @param isLabelHidden when `true` the label is not displayed but is still accessible to screen readers
 * @param isEmail when `true` input is set to type=email
 * @param isPassword when `true` type='password'. Displays visibility icon to toggle hidden on/off
 * @param isDisabled when `true` input / textarea is disabled
 * @param isOutlined when `true` the input / textarea will have border
 * @param isRequired when `true` tells browser that the input / textarea is required
 * @param autocomplete asks browser to not auto fill. Options are `new-password` or `off`
 * @param tabIndex optional tabIndex
 */
export function TextInput({
  id,
  value,
  onChange,
  label,
  placeholder,
  error,
  list,
  isLabelHidden,
  isEmail,
  isPassword,
  isDisabled,
  isOutlined,
  isTextArea,
  isRequired,
  autoComplete,
  onEnterKeyPress,
  tabIndex,
}: TextInputProps): JSX.Element {
  const [isInputHidden, setIsInputHidden] = !isPassword
    ? [undefined, undefined]
    : useState(true);

  const preventEnterKeyReset = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      e.preventDefault();
      onEnterKeyPress && onEnterKeyPress();
    }
  };

  return (
    <Container
      hasOutline={isOutlined}
      hasError={error !== undefined}
      isTextArea={isTextArea}
    >
      {setIsInputHidden !== undefined && (
        <EyeIconContainer>
          <EyeIcon
            onClick={() => setIsInputHidden(!isInputHidden)}
            isCrossedOut={isInputHidden}
          />
        </EyeIconContainer>
      )}
      {isTextArea ? (
        <TextArea
          id={id}
          value={value}
          placeholder={placeholder}
          onChange={(e) => onChange(e.target.value)}
          hasOutline={isOutlined}
          disabled={isDisabled}
          spellCheck={true}
          required={isRequired}
          tabIndex={tabIndex}
        />
      ) : (
        <Input
          id={id}
          type={isEmail ? "email" : isInputHidden ? "password" : "text"}
          value={value}
          placeholder={placeholder}
          onChange={(e) => onChange(e.target.value)}
          hasOutline={isOutlined}
          disabled={isDisabled}
          isPassword={isPassword}
          required={isRequired}
          list={list}
          onKeyPress={preventEnterKeyReset}
          autoComplete={autoComplete}
          tabIndex={tabIndex}
        />
      )}
      {isLabelHidden ? (
        <HiddenLabel htmlFor={id}>{label}</HiddenLabel>
      ) : (
        <Label hasOutline={isOutlined} isDisabled={isDisabled} htmlFor={id}>
          {label}
          {isRequired && <span>*</span>}
        </Label>
      )}
      {
        <FadeInOut
          timeout={{ appear: 600, enter: 600, exit: 600 }}
          isShowing={error !== undefined}
          render={(transitionStatus) => (
            <Error transitionStatus={transitionStatus}>{error}</Error>
          )}
        />
      }
    </Container>
  );
}

/**
 * Container wraps the entire input and controls the border and label color.
 */
const Container = styled.div<{
  hasOutline?: boolean;
  hasError?: boolean;
  isTextArea?: boolean;
}>`
  position: relative;
  width: 100%;
  height: ${(p) => (p.isTextArea ? "100px" : "auto")};
  max-height: ${(p) => (p.isTextArea ? "100px" : p.hasError ? "80px" : "65px")};
  background-color: inherit;
  transition: ${(p) => p.theme.transitions.normal};
  margin: 2px 0px;

  label {
    ${(p) => p.hasError && `color: ${p.theme.colors.error};`};
  }

  input:focus + label,
  textarea:focus + label {
    color: ${(p) => p.theme.colors.primary};
    transition: ${(p) => p.theme.transitions.normal};
  }

  input,
  textarea {
    outline: none;
    transition: ${(p) => p.theme.transitions.normal};

    border-radius: ${(p) => (p.hasOutline ? "7px" : "7px 7px 0px 0px")};
    border: ${(p) =>
      p.hasOutline
        ? `1px solid ${
            p.hasError ? p.theme.colors.error : p.theme.colors.onSurfaceLowEmp
          }`
        : "none"};
    ${(p) =>
      p.hasError &&
      !p.hasOutline &&
      `border-bottom: 1px solid ${p.theme.colors.error}`};

    :focus {
      ${(p) =>
        p.hasOutline
          ? `border: 1px solid ${p.theme.colors.primary};`
          : `border-bottom: 1px solid ${p.theme.colors.primary};`};
    }
  }

  /* Gets rid of default reveal in Edge, we have our own reveal icon and mechanism. */
  input[type="password"]::-ms-reveal,
  input[type="password"]::-ms-clear {
    display: none;
  }

  @media ${(p) => p.theme.media.tablet} {
    height: ${(p) => (p.isTextArea ? "130px" : "auto")};
    max-height: ${(p) =>
      p.isTextArea ? "100px" : p.hasError ? "85px" : "75px"};
  }
`;

const Label = styled.label<{ hasOutline?: boolean; isDisabled?: boolean }>`
  position: absolute;
  top: ${(p) => (p.hasOutline ? "1px" : "5px")};
  left: 15px;
  padding: ${(p) => (p.hasOutline ? "2px" : "0px")};
  border-radius: 2px;
  background-color: inherit;
  color: ${(p) =>
    p.isDisabled
      ? p.theme.colors.onSurfaceDisabled
      : p.theme.colors.onSurfaceLowEmp};
  font-size: 14px;

  span {
    position: relative;
    top: -2px;
    right: -2px;
  }
`;

const Input = styled.input<{ hasOutline?: boolean; isPassword?: boolean }>`
  background-color: inherit;
  color: ${(p) => p.theme.colors.onSurface};

  font-size: 16px;

  width: 100%;
  height: 50px;

  /* Increase padding right for passwords so that 'eye' doesn't cover input */
  padding: ${(p) => (p.isPassword ? "0px 37px 0px 15px" : "0px 15px")};
  margin: 10px 0px 0px 0px;

  ::placeholder {
    color: ${(p) => p.theme.colors.onSurfaceLowEmp};
    font-size: 12px;
  }

  :disabled {
    border: ${(p) =>
      p.hasOutline ? `1px solid ${p.theme.colors.onSurfaceDisabled}` : "none"};
    color: ${(p) => p.theme.colors.onSurfaceDisabled};
  }

  @media ${(p) => p.theme.media.tablet} {
    height: 60px;

    ::placeholder {
      font-size: 14px;
    }
  }
`;

const TextArea = styled.textarea<{ hasOutline?: boolean }>`
  background-color: inherit;
  color: ${(p) => p.theme.colors.onSurface};

  font-size: 16px;

  width: 100%;
  height: 83px;
  min-height: 83px;
  max-height: 83px;
  resize: none;

  padding: 15px;
  margin: 10px 0px 0px 0px;

  ::placeholder {
    color: ${(p) => p.theme.colors.onSurfaceLowEmp};
    font-size: 14px;
  }

  :disabled {
    border: ${(p) =>
      p.hasOutline ? `1px solid ${p.theme.colors.onSurfaceDisabled}` : "none"};
    color: ${(p) => p.theme.colors.onSurfaceDisabled};
  }

  @media ${(p) => p.theme.media.tablet} {
    height: 83px;
    min-height: 83px;
    max-height: 83px;
  }
`;

const Error = styled.div<{ transitionStatus: TransitionStatus }>`
  color: ${(p) => p.theme.colors.error};
  font-size: 12px;
  margin: 2px 15px 2px 15px;

  opacity: ${(p) =>
    p.transitionStatus === "entered" || p.transitionStatus === "entering"
      ? 1
      : 0};

  animation: ${FadeInKeyframe} 600ms ease-in-out;

  transition: all 600ms ease-in-out; //${(p) => p.theme.transitions.normal};

  @media ${(p) => p.theme.media.tablet} {
    font-size: 14px;
  }
`;

const EyeIconContainer = styled.div`
  position: absolute;
  right: 10px;
  top: 23px;

  @media ${(p) => p.theme.media.tablet} {
    top: 27px;
  }
`;
