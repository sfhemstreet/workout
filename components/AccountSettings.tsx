import styled from "@emotion/styled";
import React, { useState } from "react";
import {
  SwitchTransition,
  Transition,
  TransitionStatus,
} from "react-transition-group";
import { SurfaceElevation } from "../styles/SurfaceElevation";
import { User } from "../types";
import { InlineButton, SecondaryButton, SubmitButton } from "./Buttons";
import { FadeInOutMixin } from "./FadeInOut";
import {
  EMPTY_USERNAME,
  USERNAME_LENGTH,
  USERNAME_TAKEN,
} from "./SignIn/constants";
import { checkUsername } from "./SignIn/helpers";
import { TextInput } from "./TextInput";
import { H1, P } from "./Txt";

type AccountSettingsProps = {
  user: User;

  username: {
    onSubmit: (name: string) => void;
    onTypeName: (name: string) => void;
    isUnique: boolean;
    error: string | undefined;
  };

  avatar: {
    onSubmit: (url: string) => void;
    onCheckUrl: (url: string) => void;
    isUrlOk: boolean;
    error: string | undefined;
  };

  onDeleteAccount: () => void;
  onClose: () => void;
};

/**
 * AccountSettings
 * 
 * Allows signed in users to change their username / avatar, 
 * and for all user's to delete their account.
 * 
 * @param user current User
 * @param username object with properties concerning changing of username
 * @param avatar object with properties concerning changing of avatar
 * @param onDeleteAccount function called after user confirms to delete their account
 * @param onClose function called when user wishes to exit settings
 */
export const AccountSettings = ({
  user,
  username,
  avatar,
  onDeleteAccount,
  onClose,
}: AccountSettingsProps) => {
  const [newUsername, setNewUsername] = useState(user.name);
  const [usernameError, setUsernameError] = useState<string | undefined>(
    undefined
  );
  const [newAvatarUrl, setNewAvatarUrl] = useState(user.avatar);
  const [avatarError, setAvatarError] = useState<string | undefined>(undefined);

  const [isDeletingAccount, setIsDeletingAccount] = useState(false);

  const handleSubmitUsername = () => {
    if (newUsername.trim().length === 0) {
      setUsernameError(EMPTY_USERNAME);
      return;
    }

    if (newUsername.trim() === user.name) {
      setUsernameError("You are already using that name...");
      return;
    }

    if (!username.isUnique) {
      setUsernameError(USERNAME_TAKEN);
      return;
    }

    const { hasLength, hasNoSymbols, hasNoWhitespace } = checkUsername(
      newUsername.trim()
    );

    if (!hasLength) {
      setUsernameError(
        `Name must be ${USERNAME_LENGTH.min}-${USERNAME_LENGTH.max} characters.`
      );
      return;
    }
    if (!hasNoSymbols) {
      setUsernameError(`Name cannot have symbols, besides '-' and '_'`);
      return;
    }
    if (!hasNoWhitespace) {
      setUsernameError(`Name cannot contain whitespace.`);
      return;
    }

    username.onSubmit(newUsername.trim());
  };

  const handleNameChange = (name: string) => {
    setNewUsername(name);
    user.name !== name && username.onTypeName(name.trim());

    if (usernameError) setUsernameError(undefined);
  };

  const handleSubmitAvatar = () => {
    if (newAvatarUrl.trim().length === 0) {
      setAvatarError("Please enter URL to image.");
      return;
    }
    if (!avatar.isUrlOk) {
      setAvatarError("URL must be an image.");
      return;
    }
    if (newAvatarUrl.trim() === user.avatar) {
      setAvatarError("You are already using this as your avatar...");
      return;
    }

    avatar.onSubmit(newAvatarUrl.trim());
  };

  const handleAvatarChange = (url: string) => {
    setNewAvatarUrl(url);
    user.avatar !== url && avatar.onCheckUrl(url);

    if (avatarError) setAvatarError(undefined);
  };

  return (
    <SwitchTransition mode="out-in">
      <Transition
        timeout={{ enter: 0, appear: 100, exit: 300 }}
        key={isDeletingAccount ? "delete" : "edit"}
      >
        {(transitionStatus) =>
          isDeletingAccount ? (
            <Container transitionStatus={transitionStatus}>
              <CloseBtn onClick={onClose}>✕</CloseBtn>
              <H1>Delete Account</H1>
              <P>Are you sure you want to delete your account?</P>
              <Row>
                <SecondaryButton onClick={() => setIsDeletingAccount(false)}>
                  Cancel
                </SecondaryButton>
                <SubmitButton onClick={onDeleteAccount}>Delete</SubmitButton>
              </Row>
            </Container>
          ) : (
            <Container transitionStatus={transitionStatus}>
              <CloseBtn onClick={onClose}>✕</CloseBtn>
              <H1>Account Settings</H1>
              {!user.isAuthenticated && (
                <P>
                  You need to be logged in to change your username or avatar.
                </P>
              )}
              <TextInput
                id="account-settings-name"
                label="Username"
                onChange={handleNameChange}
                placeholder="Want a new username?"
                value={newUsername}
                error={
                  !username.isUnique && newUsername !== user.name
                    ? USERNAME_TAKEN
                    : usernameError
                }
                isOutlined
                isDisabled={!user.isAuthenticated}
              />
              <Row flexEnd>
                <SecondaryButton
                  onClick={handleSubmitUsername}
                  disabled={!user.isAuthenticated}
                >
                  Change Username
                </SecondaryButton>
              </Row>

              <TextInput
                id="account-settings-avatar"
                label="Avatar URL"
                onChange={handleAvatarChange}
                placeholder="Image URl"
                value={newAvatarUrl}
                error={
                  !avatar.isUrlOk
                    ? "Invalid URL, must be an image."
                    : avatarError
                }
                isOutlined
                isDisabled={!user.isAuthenticated}
              />
              <Row flexEnd>
                <SecondaryButton
                  onClick={handleSubmitAvatar}
                  disabled={!user.isAuthenticated}
                >
                  Change Avatar
                </SecondaryButton>
              </Row>
              <br />
              <Row>
                <SecondaryButton onClick={onClose}>Close</SecondaryButton>
                <SecondaryButton onClick={() => setIsDeletingAccount(true)}>
                  Delete Account
                </SecondaryButton>
              </Row>
            </Container>
          )
        }
      </Transition>
    </SwitchTransition>
  );
};

const Container = styled.article<{ transitionStatus: TransitionStatus }>`
  position: relative;
  max-width: 300px;
  width: 100%;
  height: fit-content;

  border-radius: 20px;
  ${(p) => SurfaceElevation(p.theme.name, 3)};

  padding: 1rem;

  ${(p) => FadeInOutMixin(p.transitionStatus)};

  @media ${(p) => p.theme.media.tablet} {
    max-width: 500px;
  }

  button {
    margin-top: 0.25rem;
    margin-bottom: 1rem;
  }
`;

const Row = styled.div<{ flexEnd?: boolean }>`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: ${(p) => (p.flexEnd ? "flex-end" : "space-around")};
  align-items: center;
`;

const CloseBtn = styled(InlineButton)`
  position: absolute;
  top: 2px;
  right: 12px;
  font-family: ${(p) => p.theme.font.numberFamily};
  font-size: 2rem;
  background: transparent;
`;
