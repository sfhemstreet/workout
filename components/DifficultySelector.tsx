import styled from "@emotion/styled";
import { SurfaceElevation } from "../styles/SurfaceElevation";
import { Difficulty } from "../types/Difficulty";
import { SelectableInlineButton } from "./Buttons";

const DIFFICULTY_LEVELS: Difficulty[] = ["active-rest", "moderate", "hard", "extreme"];

type DifficultySelectorProps = {
  difficulty: Difficulty;
  onSelect: (difficulty: Difficulty) => void;
};

/**
 * DifficultySelector
 * 
 * Displays all difficulty options as selectable buttons.
 * 
 * @param difficulty the selected difficulty
 * @param onSelect function called when a difficulty is selected
 */
export const DifficultySelector = ({
  difficulty,
  onSelect,
}: DifficultySelectorProps) => (
  <Row>
    {DIFFICULTY_LEVELS.map((diff) => (
      <SelectableInlineButton
        key={diff}
        isSelected={difficulty === diff}
        onClick={(e) => {
          e.preventDefault();
          onSelect(diff);
        }}
      >
        {diff}
      </SelectableInlineButton>
    ))}
  </Row>
);

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  flex-wrap: nowrap;
  margin: 10px;
  padding: 10px;

  border-radius: 10px;
  ${(p) => SurfaceElevation(p.theme.name, 3)};
`;
