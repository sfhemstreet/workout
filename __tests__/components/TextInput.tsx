import { TextInput } from "../../components/TextInput";
import { fireEvent, render } from "../test-utils/render";

test("TextInput renders properly", () => {
  const inputValues = {
    "1": "hello",
    "2": "greg@gmail.com",
    "3": "password",
    "textarea": "Text area text here"
  }
  const inputs = render(<div>
    <TextInput 
      id="1"
      value={inputValues[1]}
      onChange={(value) => inputValues[1] = value}
      label="1"
      placeholder="1"
    />
    <TextInput 
      id="2"
      value={inputValues[2]}
      onChange={(value) => inputValues[2] = value}
      label="2"
      placeholder="2"
      isEmail
    />
    <TextInput 
      id="3"
      value={inputValues[3]}
      onChange={(value) => inputValues[3] = value}
      label="3"
      placeholder="3"
      isPassword
    />
    <TextInput 
      id="outlined"
      value={""}
      onChange={() => {}}
      label="Outlined"
      placeholder="outlined"
      isOutlined
    />
    <TextInput 
      id="textarea"
      value={inputValues['textarea']}
      onChange={(value) => inputValues['textarea'] = value}
      label="TextArea"
      placeholder="TextArea"
      isTextArea
    />
  </div>);

  const input1 = inputs.getByLabelText("1") as HTMLInputElement;
  const input2 = inputs.getByLabelText("2") as HTMLInputElement;
  const input3 = inputs.getByLabelText("3") as HTMLInputElement;
  const textArea = inputs.getByLabelText("TextArea") as HTMLTextAreaElement;

  const toggleVisibilityBtn = inputs.getByRole("button");

  expect(input1.value).toBe(inputValues[1]);
  expect(input2.value).toBe(inputValues[2]);
  expect(input3.value).toBe(inputValues[3]);
  expect(textArea.value).toBe(inputValues.textarea);
  expect(input3.type).toBe("password");

  fireEvent.change(input1, { target: { value: "change 1" }});
  fireEvent.change(input2, { target: { value: "change 2" }});
  fireEvent.change(input3, { target: { value: "change 3" }});
  fireEvent.change(textArea, { target: { value: "change text area" }});
  fireEvent.click(toggleVisibilityBtn);

  expect(inputValues[1]).toBe("change 1");
  expect(inputValues[2]).toBe("change 2");
  expect(inputValues[3]).toBe("change 3");
  expect(inputValues.textarea).toBe("change text area");
  expect(input3.type).toBe("text");
  
  expect(inputs).toMatchSnapshot();
})