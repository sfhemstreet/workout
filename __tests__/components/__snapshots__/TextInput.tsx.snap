// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="css-1oijshp-Container e19wqegw5"
        >
          <input
            class="css-dn4j2n-Input e19wqegw3"
            id="1"
            placeholder="1"
            type="text"
            value="hello"
          />
          <label
            class="css-csqiz0-Label e19wqegw4"
            for="1"
          >
            1
          </label>
        </div>
        <div
          class="css-1oijshp-Container e19wqegw5"
        >
          <input
            class="css-dn4j2n-Input e19wqegw3"
            id="2"
            placeholder="2"
            type="email"
            value="greg@gmail.com"
          />
          <label
            class="css-csqiz0-Label e19wqegw4"
            for="2"
          >
            2
          </label>
        </div>
        <div
          class="css-1oijshp-Container e19wqegw5"
        >
          <div
            class="css-d4h2in-EyeIconContainer e19wqegw0"
          >
            <svg
              class="css-4i3ax9-SVGIcon e26f57s0"
              fill="none"
              height="24"
              role="button"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Visible
              </title>
              <path
                clip-rule="evenodd"
                d="M12 5C7 5 2.73 8.11 1 12.5C2.73 16.89 7 20 12 20C17 20 21.27 16.89 23 12.5C21.27 8.11 17 5 12 5ZM12 17.5C9.24 17.5 7 15.26 7 12.5C7 9.74 9.24 7.5 12 7.5C14.76 7.5 17 9.74 17 12.5C17 15.26 14.76 17.5 12 17.5ZM12 9.5C10.34 9.5 9 10.84 9 12.5C9 14.16 10.34 15.5 12 15.5C13.66 15.5 15 14.16 15 12.5C15 10.84 13.66 9.5 12 9.5Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <input
            class="css-1mjsrg1-Input e19wqegw3"
            id="3"
            placeholder="3"
            type="text"
            value="password"
          />
          <label
            class="css-csqiz0-Label e19wqegw4"
            for="3"
          >
            3
          </label>
        </div>
        <div
          class="css-y7gdxi-Container e19wqegw5"
        >
          <input
            class="css-1jng164-Input e19wqegw3"
            id="outlined"
            placeholder="outlined"
            type="text"
            value=""
          />
          <label
            class="css-1g1fvn4-Label e19wqegw4"
            for="outlined"
          >
            Outlined
          </label>
        </div>
        <div
          class="css-1ryu0wv-Container e19wqegw5"
        >
          <textarea
            class="css-1y3zcga-TextArea e19wqegw2"
            id="textarea"
            placeholder="TextArea"
            spellcheck="true"
          >
            Text area text here
          </textarea>
          <label
            class="css-csqiz0-Label e19wqegw4"
            for="textarea"
          >
            TextArea
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="css-1oijshp-Container e19wqegw5"
      >
        <input
          class="css-dn4j2n-Input e19wqegw3"
          id="1"
          placeholder="1"
          type="text"
          value="hello"
        />
        <label
          class="css-csqiz0-Label e19wqegw4"
          for="1"
        >
          1
        </label>
      </div>
      <div
        class="css-1oijshp-Container e19wqegw5"
      >
        <input
          class="css-dn4j2n-Input e19wqegw3"
          id="2"
          placeholder="2"
          type="email"
          value="greg@gmail.com"
        />
        <label
          class="css-csqiz0-Label e19wqegw4"
          for="2"
        >
          2
        </label>
      </div>
      <div
        class="css-1oijshp-Container e19wqegw5"
      >
        <div
          class="css-d4h2in-EyeIconContainer e19wqegw0"
        >
          <svg
            class="css-4i3ax9-SVGIcon e26f57s0"
            fill="none"
            height="24"
            role="button"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Visible
            </title>
            <path
              clip-rule="evenodd"
              d="M12 5C7 5 2.73 8.11 1 12.5C2.73 16.89 7 20 12 20C17 20 21.27 16.89 23 12.5C21.27 8.11 17 5 12 5ZM12 17.5C9.24 17.5 7 15.26 7 12.5C7 9.74 9.24 7.5 12 7.5C14.76 7.5 17 9.74 17 12.5C17 15.26 14.76 17.5 12 17.5ZM12 9.5C10.34 9.5 9 10.84 9 12.5C9 14.16 10.34 15.5 12 15.5C13.66 15.5 15 14.16 15 12.5C15 10.84 13.66 9.5 12 9.5Z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <input
          class="css-1mjsrg1-Input e19wqegw3"
          id="3"
          placeholder="3"
          type="text"
          value="password"
        />
        <label
          class="css-csqiz0-Label e19wqegw4"
          for="3"
        >
          3
        </label>
      </div>
      <div
        class="css-y7gdxi-Container e19wqegw5"
      >
        <input
          class="css-1jng164-Input e19wqegw3"
          id="outlined"
          placeholder="outlined"
          type="text"
          value=""
        />
        <label
          class="css-1g1fvn4-Label e19wqegw4"
          for="outlined"
        >
          Outlined
        </label>
      </div>
      <div
        class="css-1ryu0wv-Container e19wqegw5"
      >
        <textarea
          class="css-1y3zcga-TextArea e19wqegw2"
          id="textarea"
          placeholder="TextArea"
          spellcheck="true"
        >
          Text area text here
        </textarea>
        <label
          class="css-csqiz0-Label e19wqegw4"
          for="textarea"
        >
          TextArea
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
